server/coordinator/watch/watch.go

```diff
@@ -0,0 +1,154 @@
+	return &ShardWatch{
+		rootPath:       rootPath,
+		etcdClient:     client,
+		eventCallbacks: []ShardEventCallback{},
+	}
+}
+
+func (w *ShardWatch) Start(ctx context.Context) error {
+	w.lock.Lock()
+	defer w.lock.Unlock()
+
+	if w.isRunning {
+		return nil
+	}
+
+	shardsKeyPath := strings.Join([]string{w.rootPath, shardPath}, "/")
+
+	if err := w.startWatch(ctx, shardsKeyPath); err != nil {
+		return errors.WithMessage(err, "etcd register watch failed")
+	}
+
+	w.isRunning = true
+	return nil
+}
+
+func (w *ShardWatch) Stop(ctx context.Context) error {
+	w.lock.Lock()
+	defer w.lock.Unlock()
+
+	w.isRunning = false
+	w.cancel()
+	return nil
+}
+
+func (w *ShardWatch) RegisteringEventCallback(eventCallback ShardEventCallback) {
+	w.eventCallbacks = append(w.eventCallbacks, eventCallback)
+}
+
+func (w *ShardWatch) startWatch(ctx context.Context, path string) error {
+	log.Info("register shard watch", zap.String("watchPath", path))
+	go func() {
+		ctxWithCancel, cancel := context.WithCancel(ctx)
+		w.cancel = cancel
+		respChan := w.etcdClient.Watch(ctxWithCancel, path, clientv3.WithPrefix(), clientv3.WithPrevKV())
+		for resp := range respChan {
+			for _, event := range resp.Events {
+				if err := w.processEvent(event); err != nil {
+					log.Error("process event", zap.Error(err))
+				}
+			}
+		}
+	}()
+	return nil
+}
+
+func (w *ShardWatch) processEvent(event *clientv3.Event) error {
+	switch event.Type {
+	case mvccpb.DELETE:
+		shardID, err := decodeKey(string(event.Kv.Key))
+		if err != nil {
+			return err
+		}
+		oldLeader := string(event.PrevKv.Value)
+		log.Info("receive delete event", zap.String("preKV", fmt.Sprintf("%v", event.PrevKv)), zap.String("event", fmt.Sprintf("%v", event)), zap.Uint64("shardID", shardID), zap.String("oldLeader", oldLeader))
+		for _, callback := range w.eventCallbacks {
+			if err := callback.OnShardExpired(ShardExpireEvent{
+				ShardID:       storage.ShardID(shardID),
+				OldLeaderNode: oldLeader,
+			}); err != nil {
+				return err
+			}
+		}
+	case mvccpb.PUT:
+		shardID, err := decodeKey(string(event.Kv.Key))
+		if err != nil {
+			return err
+		}
+		newLeader := string(event.Kv.Value)
+		log.Info("receive put event", zap.String("event", fmt.Sprintf("%v", event)), zap.Uint64("shardID", shardID), zap.String("oldLeader", newLeader))
+		for _, callback := range w.eventCallbacks {
+			if err := callback.OnShardRegistered(ShardRegisterEvent{
+				ShardID:       storage.ShardID(shardID),
+				NewLeaderNode: newLeader,
+			}); err != nil {
+				return err
+			}
+		}
+	}
+	return nil
+}
+
+func decodeKey(keyPath string) (uint64, error) {
+	pathList := strings.Split(keyPath, "/")
+	shardID, err := strconv.ParseUint(pathList[len(pathList)-1], 10, 64)
+	if err != nil {
+		return 0, errors.WithMessage(err, "decode etcd event key failed")
+	}
+	return shardID, nil
+}
+
+func encodeKey(rootPath string, shardPath string, shardID uint64) string {
```

Comment:
```suggestion
func encodeKey(rootPath, shardPath string, shardID uint64) string {
```

BTW, this method seems used only in test codes. How about re-implementing it as `encodeShardKeyPrefix` which can be used at L67.

And rename the `decodeKey` to `decodeShardKey`.